modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxManager.java

```diff
@@ -2922,25 +3019,146 @@ public void longTransactionTimeDumpSamplesPerSecondLimit(int limit) {
      */
     public void longOperationsDumpTimeoutDistributed(long longOpsDumpTimeout) {
         broadcastToNodesSupportingFeature(
+            cctx.kernalContext(),
             new LongOperationsDumpSettingsClosure(longOpsDumpTimeout),
+            false,
             DISTRIBUTED_CHANGE_LONG_OPERATIONS_DUMP_TIMEOUT
         );
     }
 
     /**
-     * Broadcasts given job to nodes that support ignite feature.
-     *
-     * @param job Ignite job.
-     * @param feature Ignite feature.
+     * @return Collisions dump interval.
      */
-    private void broadcastToNodesSupportingFeature(IgniteRunnable job, IgniteFeatures feature) {
-        ClusterGroup grp = cctx.kernalContext().grid()
-            .cluster()
-            .forPredicate(node -> IgniteFeatures.nodeSupports(cctx.kernalContext(), node, feature));
+    public int collisionsDumpInterval() {
+        return collisionsDumpInterval;
+    }
 
-        IgniteCompute compute = cctx.kernalContext().grid().compute(grp);
+    /**
+     * @param collisionsDumpInterval New collisions dump interval or -1 for disabling.
+     */
+    public void collisionsDumpIntervalDistributed(int collisionsDumpInterval) {
+        broadcastToNodesSupportingFeature(
+            cctx.kernalContext(),
+            new TxCollisionsDumpSettingsClosure(collisionsDumpInterval),
+            true,
+            DISTRIBUTED_TX_COLLISIONS_DUMP
+        );
+    }
+
+    /** Collect queue size per key collisions info. */
+    public void pushCollidingKeysWithQueueSize(GridCacheMapEntry key, int queueSize) {
+        keyCollisionsInfo.put(key, queueSize);
+    }
 
-        compute.broadcast(job);
+    /** Wrapper for inner collect logic. */
+    private void collectTxCollisionsInfo() {
+        keyCollisionsInfo.collectInfo();
+    }
+
+    /**
+     * Check local and remote candidates queue size.
+     *
+     * @param entry CacheEntry.
+     **/
+    public void detectPossibleCollidingKeys(GridDistributedCacheEntry entry) {
+        int qSize = entry.remoteMvccSnapshot().size();
+
+        try {
+            qSize += entry.localCandidates().size();
+        }
+        catch (GridCacheEntryRemovedException ignored) {
+            // No-op, obsolete vers found.
+        }
+
+        if (qSize >= COLLISIONS_QUEUE_THRESHOLD)
+            pushCollidingKeysWithQueueSize(entry, qSize);
+    }
+
+    /** Tx key collisions info holder. */
+    private final class KeyCollisionsHolder {
+        /** Stripes count. */
+        private final int STRIPES_COUNT = cctx.kernalContext().config().getSystemThreadPoolSize();
+
+        /** Max objects per store. */
+        private static final int MAX_OBJS = 5;
+
+        /** Mutexes for each stripe. */
+        private final Object[] stripeLocks = new Object[STRIPES_COUNT];
+
+        /** Store per stripe. */
+        private final Map<GridCacheMapEntry, Integer> stores[] = new LinkedHashMap[STRIPES_COUNT];
```

Comment:
Why not use a non-blocking collection?

Reply:
Need to measure what impact it would show, in this approach i get stripes equals to current payload.